source/fuzz/transformation_permute_function_parameters.cpp

```diff
@@ -0,0 +1,179 @@
+  // Check that function exists
+  auto* function = fuzzerutil::FindFunction(context, message_.function_id());
+  if (!function ||
+      function->DefInst().opcode() != SpvOpFunction ||
+      fuzzerutil::FunctionIsEntryPoint(context, function->result_id())) {
+    return false;
+  }
+
+  if (!fuzzerutil::IsFreshId(context, message_.fresh_type_id())) {
+    return false;
+  }
+
+  // Check that permutation has valid indices
+  auto* function_type = fuzzerutil::GetFunctionType(context, function);
+  assert(function_type && "Function type is null");
+
+  const auto& permutation = message_.permutation();
+
+  // Don't take return type into account
+  uint32_t arg_size = function_type->NumInOperands() - 1;
+
+  if (static_cast<uint32_t>(permutation.size()) != arg_size) {
+    return false;
+  }
+
+  std::unordered_set<uint32_t> set;
+  for (auto index : permutation) {
+    // Don't compare an unsigned integer with 0
+    if (index >= arg_size) {
+      return false;
+    }
+
+    set.insert(index);
+  }
+
+  // Check that permutation doesn't have duplicated values
+  if (set.size() != arg_size) {
+    return false;
+  }
+
+  return true;
+}
+
+void TransformationPermuteFunctionParameters::Apply(opt::IRContext* context,
+                                              FactManager* /*unused*/) const {
+  // Retrieve all data from the message
+  uint32_t function_id = message_.function_id();
+  uint32_t fresh_type_id = message_.fresh_type_id();
+  const auto& permutation = message_.permutation();
+
+  // Find the function that will be transformed
+  auto* function = fuzzerutil::FindFunction(context, function_id);
+  assert(function && "Can't find the function");
+
+  // Find the type to transform
+  auto* function_type = fuzzerutil::GetFunctionType(context, function);
+  assert(function_type && "Function type is null");
+
+  // Create a vector of permuted arguments
+  opt::Instruction::OperandList operands = {
+    function_type->GetInOperand(0) // Function's return type
+  };
+
+  std::vector<uint32_t> operand_data = {
+    function_type->GetSingleWordInOperand(0) // Function's return type
+  };
+
+  for (auto index : permutation) {
+    // Take return type into account
+    operands.push_back(function_type->GetInOperand(index + 1));
+    operand_data.push_back(function_type->GetSingleWordInOperand(index + 1));
+  }
+
+  // Check if there is already a type with permuted arguments
+  if (uint32_t new_type = fuzzerutil::FindFunctionType(context, operand_data)) {
+    // Set function's type to that type
+    function->DefInst().SetInOperand(1, {new_type});
+  } else {
+    fuzzerutil::UpdateModuleIdBound(context, fresh_type_id);
+    // Create a new type with permuted parameters
+    auto type = MakeUnique<opt::Instruction>(
+        context, SpvOpTypeFunction, /*type_id*/ 0, fresh_type_id, operands);
+
+    context->AddType(std::move(type));
+    function->DefInst().SetInOperand(1, {fresh_type_id});
+  }
```

Comment:
Instead of having the function type be added by the transformation, could you have the transformation require that a suitable function type is present, and have the fuzzer pass apply a transformation to add the type if it is not already present?

Reply:
Fixed in fa6532d